pallets/parachain-staking/src/migrations.rs

```diff
@@ -13,3 +13,364 @@
+
+				// Must contain at least the 16 bytes of Blake2_128 hash.
+				if key_bytes.len() < 16 {
+					current = sp_io::storage::next_key(&current)?;
+					continue;
+				}
+
+				// Skip the hash and decode the AccountId.
+				key_bytes = &key_bytes[16..];
+				let mut decoder = key_bytes;
+				let maybe_collator =
+					<<T as frame_system::Config>::AccountId as Decode>::decode(&mut decoder);
+
+				if let Ok(collator) = maybe_collator {
+					// If there are no remaining bytes, then this key corresponds to the
+					// legacy single-map layout (one key per collator). If there *are*
+					// remaining bytes, it is a new double-map key which we must skip.
+					if decoder.is_empty() {
+						return Some((current.clone(), collator));
+					}
+				}
+
+				current = sp_io::storage::next_key(&current)?;
+			}
+
+			None
+		}
+
+		// Process as many legacy entries as possible within the per-step weight
+		// budget. Progress is tracked by removing legacy keys from storage and
+		// by persisting the last processed legacy key in the cursor, so the
+		// next step can resume in O(1) reads.
+		let mut used_in_step = Weight::zero();
+		let mut processed_collators: u32 = 0;
+		let mut start_from: Vec<u8> = cursor
+			.map(|c| c.to_vec())
+			.unwrap_or_else(|| prefix.to_vec());
+
+		loop {
+			let Some((full_key, collator)) = next_legacy_key::<T>(&prefix, &start_from) else {
+				// No more legacy entries to migrate – we are done. Account for
+				// the weight we actually used in this step.
+				if !used_in_step.is_zero() {
+					meter.consume(used_in_step);
+				}
+				return Ok(None);
+			};
+
+			// Decode the legacy value for this collator.
+			let Some(bytes) = sp_io::storage::get(&full_key) else {
+				// Nothing to migrate for this key; try the next one.
+				start_from = full_key;
+				continue;
+			};
+
+			let old_requests: OldScheduledRequests<T> =
+				OldScheduledRequests::<T>::decode(&mut &bytes[..]).unwrap_or_default();
+
+			let n = old_requests.len() as u64;
+			// More precise weight estimate for this specific collator based on
+			// the actual number of legacy requests `n`.
+			let reads = 1 + 3 * n;
+			let writes = 2 * n + 2;
+			let weight_for_collator = db_weight.reads_writes(reads, writes);
+
+			// Recompute remaining budget now that we know the precise weight
+			// for this collator, and ensure we do not exceed the 50% per-step
+			// safety margin.
+			let remaining_budget = step_budget.saturating_sub(used_in_step);
+			if weight_for_collator.any_gt(remaining_budget) {
+				// Cannot fit this collator into the current block's budget.
+				// Stop here and let the next step handle it.
+				break;
+			}
+
+			// Rebuild storage using the new double-map layout for this collator.
+			for request in old_requests.into_iter() {
+				let delegator = request.delegator.clone();
+
+				DelegationScheduledRequests::<T>::mutate(&collator, &delegator, |scheduled| {
+					let _ = scheduled.try_push(request);
```

Comment:
This `Error` is safe to ignore given that in the current implementation we have at most one request per collator, right?